pip/cmdoptions.py

```diff
@@ -523,6 +524,47 @@ def only_binary():
 )
 
 
+def _good_hashes():
+    """Return names of hashlib algorithms at least as strong as sha256."""
+    # Remove getattr when 2.6 dies.
+    algos = set(
+        getattr(hashlib,
+                'algorithms',
+                ('md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512')))
+    return algos - set(['md5', 'sha1', 'sha224'])
+
+
+def _merge_hash(option, opt_str, value, parser):
+    """Given a value spelled "algo:digest", append the digest to a list
+    pointed to in a dict by the algo name."""
+    if not parser.values.hashes:
+        parser.values.hashes = {}
+    try:
+        algo, digest = value.split(':', 1)
+    except ValueError:
+        parser.error('Arguments to %s must be a hash name '
+                     'followed by a value, like --hash=sha256:abcde...' %
+                     opt_str)
+    goods = _good_hashes()
+    if algo not in goods:
+        parser.error('Allowed hash algorithms for %s are %s.' %
+                     (opt_str, ', '.join(sorted(goods))))
+    parser.values.hashes.setdefault(algo, []).append(digest)
+
+
+hash = partial(
+    Option,
+    '-H', '--hash',
```

Comment:
Does this deserve a short option like ``-H``? It feels like this is something that is only rarely going to be typed out by users, especially given that you need to copy/paste (likely anyways, I doubt anyone is typing it out) a long hash digest. I'm not strictly opposed to it, but I also kind of think we should reserve giving short options for things that users are likely to actually type as part of the command invocation.